relay/blob_manager.go

```diff
@@ -0,0 +1,96 @@
+package relay
+
+import (
+	"context"
+	"fmt"
+	"github.com/Layr-Labs/eigenda/core/v2"
+	"github.com/Layr-Labs/eigenda/disperser/common/v2/blobstore"
+	"github.com/Layr-Labs/eigenda/relay/cache"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"golang.org/x/sync/errgroup"
+	"sync"
+)
+
+// blobManager encapsulates logic for fetching blobs. Utilized by the relay Server.
+// This struct adds caching and threading on top of blobstore.BlobStore.
+type blobManager struct {
+	ctx    context.Context
+	logger logging.Logger
+
+	// blobStore can be used to read blobs from S3.
+	blobStore *blobstore.BlobStore
+
+	// blobCache is an LRU cache of blobs.
+	blobCache cache.CachedAccessor[v2.BlobKey, []byte]
+
+	// pool is a work pool for managing concurrent worker goroutines.
+	pool *errgroup.Group
+}
+
+// newBlobManager creates a new blobManager.
+func newBlobManager(
+	ctx context.Context,
+	logger logging.Logger,
+	blobStore *blobstore.BlobStore,
+	blobCacheSize int,
+	workPoolSize int) (*blobManager, error) {
+
+	pool, _ := errgroup.WithContext(ctx)
+	pool.SetLimit(workPoolSize)
+
+	server := &blobManager{
+		ctx:       ctx,
+		logger:    logger,
+		blobStore: blobStore,
+		pool:      pool,
+	}
+
+	cache, err := cache.NewCachedAccessor[v2.BlobKey, []byte](blobCacheSize, server.fetchBlob)
+	if err != nil {
+		return nil, fmt.Errorf("error creating blob cache: %w", err)
+	}
+	server.blobCache = cache
+
+	return server, nil
+}
+
+// GetBlob retrieves a blob from the blob store.
+func (s *blobManager) GetBlob(blobKey v2.BlobKey) ([]byte, error) {
+
+	// Even though we don't need extra parallelism here, we still use the work pool to ensure that we don't
+	// permit too many concurrent requests to the blob store.
```

Comment:
What happens if there are too many concurrent requests to S3? I'm wondering if it could be worth using this blob manager on the encoder (although it doesn't need the caching part).

See: https://github.com/Layr-Labs/eigenda/blob/94ab42f939b12d7e4ea3957aa488a460ef5d76d0/disperser/encoder/server_v2.go#L128

Reply:
Nothing catastrophic if there is too much S3 traffic, other than a potential slowdown. After a certain point, adding more goroutines attempting to pull data will result in a net reduction in capacity. Think of this sort of code as a defensive programming in case the node is overloaded with work.